fix(provided-externals): logs warning for child modules

Francois-Esquire opened the pull request:
Replaced error thrown `onModuleLoad` with warning log for child modules that contain the `providedExternals` in `appConfig`.

10xLaCroixDrinker commented on prod-sample/sample-modules/late-frank/0.0.1/src/index.js:
```diff
@@ -0,0 +1,25 @@
+import ModuleContainer from './components/LateFrank';
+
+if (!global.BROWSER) {
+  ModuleContainer.appConfig = {
```
this should be done by the bundler

Francois-Esquire replied:
What do you mean by this specifically? The configuration?

Francois-Esquire replied:
I see what you mean now, will update.

Francois-Esquire replied:
updated aa0b3a1

Francois-Esquire commented:
☑️ Preflight Checklist:

What is the communication plan for this change?
None, `providedExternals` pending documentation

Does any documentation need to be updated or added to account for this change? If so was it done already?
No

What is the motivation for this change?
To prevent the app from crashing based on child module mis-configuration

Should these changes also be applied to a maintenance branch or any other long lived branch?
No

How was this change tested?
Unit and integration testing

Does this change require cross browser checks? Why or why not?
No, server side behavior

Does this change require a performance test prior to merging? Why or why not?
No, replaced throwing an error with `console.warn`

Could this be considered a breaking change? Why or why not?
No, API is unchanged

Does the change impact caching?
No

Does the change impact HTTP headers?
No

Does the change have any new infrastructure requirements?
No

Does the change affect other versions of the app?
No

Does the change require additional environment variables?
No

What is the impact to tenants?
None

What is the impact to individual users?
Warning on mis-configuration of `appConfig`

What is the change to the size of assets?
None

Should integration tests be added to protect against future regressions on this change?
Added

10xLaCroixDrinker commented on __tests__/integration/one-app.spec.js:
```diff
@@ -568,6 +568,41 @@ describe('Tests that require Docker setup', () => {
+        });
+
+        test(
+          'writes a warning to log if a child module has set `providedExternals` in `appConfig`',
```
you need to test that the externals provided by the root are still correct/available and that the ones provided by the non-root module are not available

Francois-Esquire replied:
Got it, I'll update the tests

Francois-Esquire replied:
updated aa0b3a1

10xLaCroixDrinker commented on src/server/utils/onModuleLoad.js:
```diff
@@ -118,7 +118,9 @@ export default function onModuleLoad({
-    throw new Error(`Module ${moduleName} attempted to provide externals, but it is not the root module.`);
+    console.warn(
+      `Module ${moduleName} attempted to provide externals. Only the root module can provide externals.`
+    );
```
You will need some changes around here: https://github.com/americanexpress/one-app-cli/blob/master/packages/one-app-bundler/webpack/loaders/provided-externals-loader.js#L34

Francois-Esquire replied:
I think it's best we move that helper added by the loader into One App -> `getRootModule` and remove it from being injected. Thoughts?

mtomcal replied:
- [ ] Get @10xLaCroixDrinker Feedback on this
